javascript/computer_science/hash_map_data_structure.md

```diff
@@ -0,0 +1,204 @@
+
+You might have spotted a problem: what if our school is populated with many people whose names share the same first letter `C`? Then we will have a directory labeled `C` that holds too many names while other directories could be empty. To eliminate this duplication and better separate our students, we need to rework our hash function.
+
+```javascript
+function hash(name, surname) {
+  return name.charAt(0) + surname.charAt(0);
+}
+```
+
+Instead of just taking the first name letter, we take the first name and last name letters. `"Carlos Smith"` will have a hash code of `"CS"`. This will spread our students among more directories and will eliminate many duplicate hash codes from being generated.
+
+![Example of hashing a name using first name's first letter, and last name's first letter](./hash_map_data_structure/imgs/00.png)
+
+But it still doesn't solve our problem. What if we have a common combination of first letters in students' names? Then we will still have an imbalance in the size of the directories. We need to make it easier to find the person we're looking for, so let's rework our hash code.
+
+```javascript
+function stringToNumber(string) {
+  let hashCode = 0;
+  for (let i = 0; i < string.length; i++) {
+    hashCode += string.charCodeAt(i);
+  }
+
+  return hashCode;
+}
+
+function hash(name, surname) {
+  return stringToNumber(name) + stringToNumber(surname);
+}
+```
+
+We not only consider the first letters with this technique. Instead, we take the entire name and convert it into numbers.
+
+You might be thinking, wouldn't it be just better to save the whole name as a hash code? That is true. This would make it unique for each name, but in the context of hash maps, we need hash code to be a number. This number will serve as the index to the bucket that will store the key value pair. More on buckets in the next section.
+
+### Buckets
+
+Buckets are storage that we need to store our elements. Simply, it's an array. For a specific key, we decide which bucket to use for storage through our hash function. Hash function returns a number which serves as the index of the array at which we store this specific key value pair. Let's say we need to store a key "Fred":
+
+1. Pass "Fred" into the hash function to get the hash code which is `508`
+1. Find the bucket at index `508`
+1. Store the key value pair in that bucket.
+
+This is an oversimplified explanation; we'll discuss more internal mechanics later in the lesson.
+
+To get a value using a key, we put each entry inside a bucket as a `Node` item, which hold both the key and the value. To retrieve the value we hash a key, we find the bucket number, if the bucket is not empty then we go to that bucket we compare if the Node's key is the same key that is already in the bucket, if it is then we retrieve the Node's value otherwise we return null.
```

Comment:
```suggestion
To get a value using a key, we put each entry inside a bucket as a `Node` item, which holds both the key and the value. To retrieve the value, we hash the key and calculate its bucket number. If the bucket is not empty, then we go to that bucket and compare if the Node's key is the same key that is already in the bucket. If it is, then we can return the Node's value. Otherwise, we return null.
```